compiler/rustc_incremental/src/persist/load.rs

```diff
@@ -181,14 +190,33 @@ pub fn load_dep_graph(sess: &Session) -> DepGraphFuture {
                     // We can't reuse the cache, purge it.
                     debug!("load_dep_graph_new: differing commandline arg hashes");
 
+                    // If we've asserted that an existing cache was loaded,
+                    // compilation should end, since we just ignored the cache.
+                    if incr_state_assertion == IncrementalStateAssertion::Loaded {
```

Comment:
This is the only place that you check for `IncrementalStateAssertion::Loaded`. However, if we fail to load the data entirely (e.g. we take the `LoadResult::Error` match arm above), then we will never hit this check.

I think a simpler way to perform the check would be to move all of this logic to `LoadResult::Open`: https://github.com/rust-lang/rust/blob/c7e4740ec18996e082fe6e29ebf7efdc7dda418f/compiler/rustc_incremental/src/persist/load.rs#L26-L45

At that point, the final load result is known. To make things simpler, you could change the command-line arguments to be a simple "loaded" / "not-loaded", ignoring the exact reason why the failure happened. If more detail turns out to be useful for certain incremental tests, the options could be modified in a follow-up PR (this is a perma-unstable option, so we don't need to worry about making breaking changes).

If you have any questions, feel free to message me on Zulip: https://rust-lang.zulipchat.com/